### PR TITLE
New functionality and code improvements

### PR DESCRIPTION
**Pull request by shadre**
Open for feedback.

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  printf("%-7s", "| #{prod[:qty]}")
+  printf("%-15s", "| #{'%.02f' % prod[:price]} EUR")
+  puts
+  end
+  puts " ==============================================================="
+  printf("%64s", "Total SUM: #{basket.sum} EUR")
+  puts
+  printf("%64s", "Total SUM with VAT: #{basket.sum_with_vat} EUR")
+  puts
+
```
Comment (Kacper3331) [inline main.rb:81]: You have two main.rb files one inside `\lib` and one outside.

File: lib/basket.rb [hunk 2]
```diff
@@ -1,44 +1,44 @@
...
 
   def sum
-    @products.
-    map(&:price).
-    reduce(0, :+)    
+    sum = 0
+    summary.each do |prod|
+      sum += prod[:price]
+    end
+    sum.round(2)
```
Comment (smt116) [inline lib/basket.rb]: It doesn't look good. Maybe there is a better solution in Ruby?

File: spec/basket_spec.rb [hunk 3]
```diff
@@ -2,19 +2,60 @@
...
 RSpec.describe Basket do
 
-  it "creates basket" do
-    expect Basket.new.is_a? Basket
+  let(:football) { Product.new(name: "Football", price: 79) }
+  let(:basketball) { Product.new(name: "Basketball", price: 60) }
+  let(:tennis_ball) { Product.new(name: "Tennis ball", price: 10) }
+  let(:footballs_in_basket) { BasketItem.new(product: football, quantity: 2)}
+  let(:basketballs_in_basket) { BasketItem.new(product: basketball, quantity: 1)}
+  let(:tennis_balls_in_basket) { BasketItem.new(product: tennis_ball, quantity: 3)}
```
Comment (Kacper3331) [inline spec/basket_spec.rb]: Space before the closing bracket in lines 8-10.

File: lib/basketitem.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    raise ArgumentError unless num.is_a?(Integer)
+
+    new_quantity = @quantity + num
+
+    unless new_quantity < 0
+    ##(it's OK to set @quantity = 0 here, because Basket .summary method rejects such items)
+      @quantity = new_quantity
+    else
+      raise WrongQuantityRequested
+    end
```
Comment (smt116) [inline lib/basketitem.rb:24]: It looks weird that you have new variable that is assigned to the old one only when some condition is fulfilled. Maybe you can check that condition "in fly"?

Another thing is your comment. Why there are two hash characters on the beginning? Why there are brackets? Why there is no spacebar after hash? ...

imho `unless` should not be used with `else`. Use `if - else` with proper condition instead.

File: spec/basket_spec.rb [hunk 5]
```diff
@@ -2,19 +2,60 @@
...
-  it "checks if sum with vat is correct" do
-    sum = Basket.new([p1, p2]).sum
-    sum_with_vat = Basket.new([p1, p2]).sum_with_vat
-    expect(sum * 1.23).to eq sum_with_vat
+  describe "#add" do
+    it "adds selected item to basket" do
+      basket.add(tennis_balls_in_basket)
+        products_in_basket_names = basket.items.map do |item|
+         { name: item.product.name }
+        end
```
Comment (smt116) [inline spec/basket_spec.rb]: indentation

File: spec/basket_spec.rb [hunk 6]
```diff
@@ -2,19 +2,60 @@
...
+      expect(products_in_basket_names.select { |prod| prod[:name] == "Tennis ball" }.empty?).to eq(false)
```
Comment (smt116) [inline spec/basket_spec.rb]: assigne expect's body to variable, put it before and use it here.

File: spec/basketitem_spec.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
+require_relative '../lib/basketitem.rb'
+
+RSpec.describe BasketItem do
+
+  let(:football) { Product.new(name: "Football", price: 79) }
+  let(:basketball) { Product.new(name: "Basketball", price: 60) }
+  let(:tennis_ball) { Product.new(name: "Tennis ball", price: 10) }
+
+  describe "#new" do
+    it "returns argument error when product not provided" do
```
Comment (smt116) [inline spec/basketitem_spec.rb]: it will not return error. It will raise an exception / error

File: spec/basketitem_spec.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
+require_relative '../lib/basketitem.rb'
+
+RSpec.describe BasketItem do
+
```
Comment (smt116) [inline spec/basketitem_spec.rb]: remove empty line

File: main.rb (hunk 1 above)
Comment (shadre) [inline main.rb:81]: True. This was done by mistake. Thanks for spotting it! :)

File: spec/basket_spec.rb (hunk 6 above)
Comment (shadre) [inline spec/basket_spec.rb]: Will do. Just to know for future: is this for readability?
If yes, is there any rule of thumb for when to do it?
Or just common sense? :)

Comment (smt116) [inline spec/basket_spec.rb]: Yes, it is for readability. There is no golden rule for all cases. You need to be able to recognize what is happening on the first glance.

File: lib/basket.rb [hunk 9]
```diff
@@ -31,11 +31,10 @@ def summary
...
   def sum
-    sum = 0
-    summary.each do |prod|
-      sum += prod[:price]
-    end
-    sum.round(2)
+    summary
+    .map{|prod| prod[:price]}
+    .reduce(:+)
+    .round(2)
```
Comment (smt116) [inline lib/basket.rb]: wrong indentation

File: lib/basket.rb [hunk 10]
```diff
@@ -31,11 +31,10 @@ def summary
   end
 
   def sum
-    sum = 0
-    summary.each do |prod|
-      sum += prod[:price]
-    end
-    sum.round(2)
+    summary
+    .map{|prod| prod[:price]}
```
Comment (smt116) [inline lib/basket.rb]: missing spacebars around curly brackets

File: lib/basketitem.rb [hunk 11]
```diff
@@ -16,17 +16,15 @@ def initialize(product:, quantity: 1)
...
     raise ArgumentError unless num.is_a?(Integer)
 
-    new_quantity = @quantity + num
-
-    unless new_quantity < 0
-    ##(it's OK to set @quantity = 0 here, because Basket .summary method rejects such items)
-      @quantity = new_quantity
+    if @quantity + num >= 0
+    # it's OK to set @quantity = 0 here, because Basket#summary method rejects such items
+      @quantity = @quantity + num
```
Comment (smt116) [inline lib/basketitem.rb]: `@quantity += num`

File: lib/basketitem.rb [hunk 12]
```diff
@@ -16,17 +16,15 @@ def initialize(product:, quantity: 1)
   def change_quantity(num)
     raise ArgumentError unless num.is_a?(Integer)
 
-    new_quantity = @quantity + num
-
-    unless new_quantity < 0
-    ##(it's OK to set @quantity = 0 here, because Basket .summary method rejects such items)
-      @quantity = new_quantity
+    if @quantity + num >= 0
+    # it's OK to set @quantity = 0 here, because Basket#summary method rejects such items
```
Comment (smt116) [inline lib/basketitem.rb]: indentation for comment as well ;)

File: spec/basketitem_spec.rb [hunk 13]
```diff
@@ -1,13 +1,12 @@
...
 
 RSpec.describe BasketItem do
-
   let(:football) { Product.new(name: "Football", price: 79) }
   let(:basketball) { Product.new(name: "Basketball", price: 60) }
   let(:tennis_ball) { Product.new(name: "Tennis ball", price: 10) }
 
   describe "#new" do
-    it "returns argument error when product not provided" do
+    it "raises argument error when product not provided" do
```
Comment (smt116) [inline spec/basketitem_spec.rb:9]: 👍

File: lib/basket.rb [hunk 14]
```diff
@@ -32,7 +32,7 @@ def summary
 
   def sum
     summary
-    .map{|prod| prod[:price]}
+    .map{ |prod| prod[:price] }
```
Comment (smt116) [inline lib/basket.rb]: spacebar also before `{`

**Pull request merged**